crates/core/src/operations/add_column.rs

```diff
@@ -0,0 +1,125 @@
+//! Add a new column to a table
+
+use futures::future::BoxFuture;
+use itertools::Itertools;
+use maplit::hashset;
+
+use super::cast::merge_struct;
+use super::transaction::{CommitBuilder, CommitProperties, PROTOCOL};
+
+use crate::kernel::{Protocol, ReaderFeatures, StructField, WriterFeatures};
+use crate::logstore::LogStoreRef;
+use crate::operations::set_tbl_properties::convert_properties_to_features;
+use crate::protocol::DeltaOperation;
+use crate::table::state::DeltaTableState;
+use crate::{DeltaResult, DeltaTable, DeltaTableError};
+
+/// Add new columns and/or nested fields to a table
+pub struct AddColumnBuilder {
+    /// A snapshot of the table's state
+    snapshot: DeltaTableState,
+    /// Fields to add/merge into schema
+    fields: Option<Vec<StructField>>,
+    /// Delta object store for handling data files
+    log_store: LogStoreRef,
+    /// Additional information to add to the commit
+    commit_properties: CommitProperties,
+}
+
+impl super::Operation<()> for AddColumnBuilder {}
+
+impl AddColumnBuilder {
+    /// Create a new builder
+    pub fn new(log_store: LogStoreRef, snapshot: DeltaTableState) -> Self {
+        Self {
+            snapshot,
+            log_store,
+            fields: None,
+            commit_properties: CommitProperties::default(),
+        }
+    }
+
+    /// Specify the fields to be added
+    pub fn with_fields(mut self, fields: Vec<StructField>) -> Self {
+        self.fields = Some(fields);
+        self
+    }
+    /// Additional metadata to be added to commit info
+    pub fn with_commit_properties(mut self, commit_properties: CommitProperties) -> Self {
+        self.commit_properties = commit_properties;
+        self
+    }
+}
+
+impl std::future::IntoFuture for AddColumnBuilder {
+    type Output = DeltaResult<DeltaTable>;
+
+    type IntoFuture = BoxFuture<'static, Self::Output>;
+
+    fn into_future(self) -> Self::IntoFuture {
+        let this = self;
+
+        Box::pin(async move {
+            let mut metadata = this.snapshot.metadata().clone();
+            let fields = match this.fields {
+                Some(v) => v,
+                None => return Err(DeltaTableError::Generic("No fields provided".to_string())),
+            };
+
+            let table_schema = this.snapshot.schema();
+            let new_table_schema =
+                merge_struct(table_schema, &fields.clone().into_iter().collect())?;
+
+            // TODO(ion): Think of a way how we can simply this checking through the API or centralize some checks.
+            let contains_timestampntz = PROTOCOL.contains_timestampntz(&fields);
+            let protocol = this.snapshot.protocol();
+
+            let maybe_new_protocol = if contains_timestampntz {
+                if protocol.min_reader_version == 3 && protocol.min_writer_version == 7 {
```

Comment:
So if protocol is 3,7 you enable timestamp ntz, otherwise you set the protocol to 3,7 and enable timezone ntz? Am I missing something or does this seem redundant? Wouldn't you just check if it has timestamps and update the protocol/add table features? It seems unnecessary to create a new one.

Reply:
Afaik, if we just add a new value to the HashSet of table features, it would would not automatically bump also reader version 3 and writer version 7, perhaps that should be implemented to make this checks redundant.

So if you add a table feature, just modify itself to bump reader and writer.